doxygen/man/man3/GUIUtil.3
.TH "GUIUtil" 3 "Wed Feb 10 2016" "Version 1.0.0.0" "darksilk" \" -*- nroff -*-
.ad l
.nh
.SH NAME
GUIUtil \- 
.SH SYNOPSIS
.br
.PP
.SS "Classes"

.in +1c
.ti -1c
.RI "class \fBClickableLabel\fP"
.br
.ti -1c
.RI "class \fBHelpMessageBox\fP"
.br
.ti -1c
.RI "class \fBToolTipToRichTextFilter\fP"
.br
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "QString \fBdateTimeStr\fP (const QDateTime &date)"
.br
.ti -1c
.RI "QString \fBdateTimeStr\fP (qint64 nTime)"
.br
.ti -1c
.RI "QFont \fBdarksilkAddressFont\fP ()"
.br
.ti -1c
.RI "void \fBsetupAddressWidget\fP (QLineEdit *widget, QWidget *parent)"
.br
.ti -1c
.RI "void \fBsetupAmountWidget\fP (QLineEdit *widget, QWidget *parent)"
.br
.ti -1c
.RI "bool \fBparseDarkSilkURI\fP (const QUrl &uri, \fBSendCoinsRecipient\fP *out)"
.br
.ti -1c
.RI "bool \fBparseDarkSilkURI\fP (QString uri, \fBSendCoinsRecipient\fP *out)"
.br
.ti -1c
.RI "QString \fBHtmlEscape\fP (const QString &str, bool fMultiLine)"
.br
.ti -1c
.RI "QString \fBHtmlEscape\fP (const std::string &str, bool fMultiLine)"
.br
.ti -1c
.RI "void \fBcopyEntryData\fP (QAbstractItemView *view, int column, int role)"
.br
.ti -1c
.RI "QString \fBgetSaveFileName\fP (QWidget *parent, const QString &caption, const QString &dir, const QString &filter, QString *selectedSuffixOut)"
.br
.ti -1c
.RI "Qt::ConnectionType \fBblockingGUIThreadConnection\fP ()"
.br
.ti -1c
.RI "bool \fBcheckPoint\fP (const QPoint &p, const QWidget *w)"
.br
.ti -1c
.RI "bool \fBisObscured\fP (QWidget *w)"
.br
.ti -1c
.RI "void \fBopenDebugLogfile\fP ()"
.br
.ti -1c
.RI "bool \fBGetStartOnSystemStartup\fP ()"
.br
.ti -1c
.RI "bool \fBSetStartOnSystemStartup\fP (bool fAutoStart)"
.br
.ti -1c
.RI "void \fBSetBlackThemeQSS\fP (QApplication &app)"
.br
.ti -1c
.RI "QString \fBformatPingTime\fP (double dPingTime)"
.br
.ti -1c
.RI "QString \fBformatDurationStr\fP (int secs)"
.br
.ti -1c
.RI "QString \fBformatServicesStr\fP (quint64 mask)"
.br
.in -1c
.SH "Detailed Description"
.PP 
Utility functions used by the DarkSilk Qt UI\&. 
.SH "Function Documentation"
.PP 
.SS "Qt::ConnectionType GUIUtil::blockingGUIThreadConnection ()"
Get connection type to call object slot in GUI thread with invokeMethod\&. The call will be blocking\&.
.PP
\fBReturns:\fP
.RS 4
If called from the GUI thread, return a Qt::DirectConnection\&. If called from another thread, return a Qt::BlockingQueuedConnection\&. 
.RE
.PP

.PP
Definition at line 220 of file guiutil\&.cpp\&.
.SS "bool GUIUtil::checkPoint (const QPoint & p, const QWidget * w)"

.PP
Definition at line 232 of file guiutil\&.cpp\&.
.SS "void GUIUtil::copyEntryData (QAbstractItemView * view, int column, int role = \fCQt::EditRole\fP)"
Copy a field of the currently selected entry of a view to the clipboard\&. Does nothing if nothing is selected\&. 
.PP
\fBParameters:\fP
.RS 4
\fIcolumn\fP Data column to extract from the model 
.br
\fIrole\fP Data role to extract from the model 
.RE
.PP
\fBSee also:\fP
.RS 4
TransactionView::copyLabel, TransactionView::copyAmount, TransactionView::copyAddress 
.RE
.PP

.PP
Definition at line 161 of file guiutil\&.cpp\&.
.SS "QFont GUIUtil::darksilkAddressFont ()"

.PP
Definition at line 58 of file guiutil\&.cpp\&.
.SS "QString GUIUtil::dateTimeStr (const QDateTime & date)"

.PP
Definition at line 48 of file guiutil\&.cpp\&.
.SS "QString GUIUtil::dateTimeStr (qint64 nTime)"

.PP
Definition at line 53 of file guiutil\&.cpp\&.
.SS "QString GUIUtil::formatDurationStr (int secs)"

.PP
Definition at line 524 of file guiutil\&.cpp\&.
.SS "QString GUIUtil::formatPingTime (double dPingTime)"

.PP
Definition at line 519 of file guiutil\&.cpp\&.
.SS "QString GUIUtil::formatServicesStr (quint64 mask)"

.PP
Definition at line 544 of file guiutil\&.cpp\&.
.SS "QString GUIUtil::getSaveFileName (QWidget * parent = \fC0\fP, const QString & caption = \fCQString()\fP, const QString & dir = \fCQString()\fP, const QString & filter = \fCQString()\fP, QString * selectedSuffixOut = \fC0\fP)"
Get save filename, mimics QFileDialog::getSaveFileName, except that it appends a default suffix when no suffix is provided by the user\&.
.PP
\fBParameters:\fP
.RS 4
\fIparent\fP Parent window (or 0) 
.br
\fIcaption\fP Window caption (or empty, for default) 
.br
\fIdir\fP Starting directory (or empty, to default to documents directory) 
.br
\fIfilter\fP Filter specification such as 'Comma Separated Files (*\&.csv)' 
.br
\fIselectedSuffixOut\fP Pointer to return the suffix (file type) that was selected (or 0)\&. Can be useful when choosing the save file format based on suffix\&. 
.RE
.PP

.PP
Definition at line 174 of file guiutil\&.cpp\&.
.SS "bool GUIUtil::GetStartOnSystemStartup ()"

.PP
Definition at line 420 of file guiutil\&.cpp\&.
.SS "QString GUIUtil::HtmlEscape (const QString & str, bool fMultiLine)"

.PP
Definition at line 145 of file guiutil\&.cpp\&.
.SS "QString GUIUtil::HtmlEscape (const std::string & str, bool fMultiLine)"

.PP
Definition at line 156 of file guiutil\&.cpp\&.
.SS "bool GUIUtil::isObscured (QWidget * w)"

.PP
Definition at line 239 of file guiutil\&.cpp\&.
.SS "void GUIUtil::openDebugLogfile ()"

.PP
Definition at line 248 of file guiutil\&.cpp\&.
.SS "bool GUIUtil::parseDarkSilkURI (const QUrl & uri, \fBSendCoinsRecipient\fP * out)"

.PP
Definition at line 85 of file guiutil\&.cpp\&.
.SS "bool GUIUtil::parseDarkSilkURI (QString uri, \fBSendCoinsRecipient\fP * out)"

.PP
Definition at line 131 of file guiutil\&.cpp\&.
.SS "void GUIUtil::SetBlackThemeQSS (QApplication & app)"

.PP
Definition at line 481 of file guiutil\&.cpp\&.
.SS "bool GUIUtil::SetStartOnSystemStartup (bool fAutoStart)"

.PP
Definition at line 421 of file guiutil\&.cpp\&.
.SS "void GUIUtil::setupAddressWidget (QLineEdit * widget, QWidget * parent)"

.PP
Definition at line 69 of file guiutil\&.cpp\&.
.SS "void GUIUtil::setupAmountWidget (QLineEdit * widget, QWidget * parent)"

.PP
Definition at line 76 of file guiutil\&.cpp\&.
.SH "Author"
.PP 
Generated automatically by Doxygen for darksilk from the source code\&.


CHANGELOG
Base58 Change, Port Changes and Minor Fixes
Addition of OBJ directory for Daemon builds
Version updated
ReadMe updated
Instamine code for Silk/Dark weaver swap/burn added to main.cpp
OpenSSL update
LevelDB 1.18 Update
Fix to guard against OpenSSL's new strict DER checks, fixed getbalance and other minor fixes
Spelling edits
Enabled locking of Stormnodes via the Stormnode configuration file
Unset Sandstorm checkbox if there were non-anonymized inputs selected in coincontrol, give warning
Removal of target space change and hardfork
Fixed time to switch "out of sync" messages
Fix twice denominatedBalance calculations
Solve performance problems by caching and fix probable security issue
RPC: Allow "sandstorm" only with ENABLE_WALLET
Disconnect Sandstorm timer on Overview page destruction
Fix spaces/monospace fonts in qt console
Mac compatibility so users can switch tabs using CMD+(NUM)
Added better instructions to get libsecp256k1 installed in case of library issues within ReadMe
Added secp256k1 libs to /deps in case for any reason the libs are not available from github anymore
Increased maximum block size to 50MB (50x Bitcoin Core)
Added safe shutdown for Windows
Fixed sync issues and CPU usage
Add event support for Tor
rpc: Prevent easy memory exhaustion attack 
Fixed issue where using setaccount with a foreign address causes the address to be added to your receiving addresses.
Fix priority calculation in CreateTransaction
darksilk.conf added to /src
GUI Styling changes
Statistics Page Added and Icons Updated
Clickable Padlock to Unlock Wallet and Unlock for Staking Only
Fixed GUI bugs
Assertion Error Fix for MacOSX
Sandstorm GUI Fixes
Decentralised DarkSilkMarket added
Overview Transactions UI Fixes
Messaging, GUI Update, Null Pointer Fix, Various Fixes/Changes
Windows Build Fix
Addition of possibly needed dependency
Stormnode Collateral Increase
Marketplace RPC Calls | Cleanup
Amended Spork Max Value for Stormnode Collateral
Removal of unused Spork
Minimum Transaction fee changed to 0.0001DRKSLK
Add support for miniupnpc API version 14
DARKSILK_TIMEDATA_MAX_SAMPLES 200
Add time offset to getpeerinfo output
Don't check duplicate stake while importing
Ignore blocks received while importing
Further Market RPC JSON calls
marketbuyrequest RPC command
marketmybuys RPC call for market
MacOSX Fixes and GUI Fix
Market RPC Minor Fixes
darksilk-qt.pro cleanup
DarkSilk-qt.app renamed to DarkSilk.app for OSX
Replaced stake min age with min confirmation
Limited drop of difficulty per block
Made special marker non spendable
Added nonspendable encoding for block signing key
BIP66 compatibility
Replaced NOP7 with CHECKLOCKTIMEVERIFY
Enforce new script verification flags
Static pos reward
Demand canonical block sig from updated peers
Submitblock: Add submission of stake template
Reward and Min Conf Blocks Changed
Checkpointing Removal
Gitian MacDeploy Graphics Added
IsProtocolV1/IsProtocolV2 hard fork switches removed
FutureDriftV1/FutureDriftV2 reverted back to FutureDrift
CheckStakeKernelHashV1/CheckStakeKernalHashV2 reverted back to CheckStakeKernelHash
Unsigned Int GetTargetSpacing changed to Static Const Int POS_TARGET_SPACING
Fix boost::get usage with Boost 1.58 in rpcrawtransaction.cpp 
Static PoS Reward of 1 DRKSLK
Prevent redefinition compiler warning
Fix issue where getrawtransaction RPC call does not work
PoW additions to GetNextTargetRequired
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
LevelDB windows cross compile missing files
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
Removal of Dead Code
Fix for building with Boost 1.58
LevelDB Cross compile fixes
DarkSilk-Qt: cleanup / optimise addressbookpage
Discourage fee sniping with nLockTime
Removal of dead code from main.h
Changed ProofOfStakeReward to use static const
Diff changes
Shadowchat Messaging Updates
DNSSeed Thread Fix
Stormnode payments start on block 82002
Restore hex to getrawtransaction vout scriptPubkey
static const rename | nActualSpacing code reform
Removal of Keepass until later date that this doesn't cause an issue with Windows builds
Updated dependencies, added Gitian OSX files
Remove unused GetProofOfStakeReward
Faster Retargetting for PoW
Fixed PoS target spacing
Fix payee selectoin on StormnodePayments::ProcessBlock
fix log output / more notCapableReason-s
HasCollateralInputs should check only for at least 1 collateral not 2
Track lastTimeChanged for mnodeman and ask for update only if UpdateNeeded
Added sstx support for free transactions 
Fix for src/qt/stormnodemanager.cpp
RPC Fixes
Fix DoAutomaticDenominating
Few Fixes for Processing Extra Messages
Reversioning and restriction change
Check and remove expired Stormnodes on Client start
Maps in snodeman
bool warning fix
SSEG Update
Show new number of SN's on add/remove in logs | Move output to debug level
sigTime
Temp Fix prior to major changes
SPORK Additions/Changes | SANDSTORM fee removal
prevent ganing sigTime too much on initial ssee
SandStorm/InstantX Functionality | Send with SandStorm on sendcoinsdialog.ui
Fix Filtering for active | add spaces to table-fy console view | add spaces to IP only in full mode
Merged Stormnode blinding
fix compiling for windows
fix console spam
Make stormnodes.dat verification more specific, prevent from overwriting corruped
files with unknown format, give more info in debug.log output
fixed shared key usage
fix even more console spam
rename stormnodes.dat->mncache.dat / fix strings
save nsqCount in mncache.dat
[Qt] add more NULL pointer checks in darksilkgui.cpp
close wallet regardless of "minimize to tray" settings
Added coincontrol sandstorm rounds column
Stormnode Proof of Service Added

src/qt/guiutil.cpp
#include <QApplication>

#include "guiutil.h"
#include "net.h"
#include "protocol.h"
#include "chainparams.h"
#include "darksilkaddressvalidator.h"
#include "walletmodel.h"
#include "darksilkunits.h"

#include "util.h"
#include "init.h"

#include <QDateTime>
#include <QDoubleValidator>
#include <QFont>
#include <QLineEdit>
#include <QUrl>
#include <QTextDocument> // For Qt::escape
#include <QAbstractItemView>
#include <QClipboard>
#include <QFileDialog>
#include <QDesktopServices>
#include <QThread>

#include <boost/filesystem.hpp>
#include <boost/filesystem/fstream.hpp>

#ifdef WIN32
#ifdef _WIN32_WINNT
#undef _WIN32_WINNT
#endif
#define _WIN32_WINNT 0x0501
#ifdef _WIN32_IE
#undef _WIN32_IE
#endif
#define _WIN32_IE 0x0501
#define WIN32_LEAN_AND_MEAN 1
#ifndef NOMINMAX
#define NOMINMAX
#endif
#include "shlwapi.h"
#include "shlobj.h"
#include "shellapi.h"
#endif

namespace GUIUtil {

QString dateTimeStr(const QDateTime &date)
{
    return date.date().toString(Qt::SystemLocaleShortDate) + QString(" ") + date.toString("hh:mm");
}

QString dateTimeStr(qint64 nTime)
{
    return dateTimeStr(QDateTime::fromTime_t((qint32)nTime));
}

QFont darksilkAddressFont()
{
    QFont font("Monospace");
#if QT_VERSION >= 0x040800
    font.setStyleHint(QFont::Monospace);
#else
    font.setStyleHint(QFont::TypeWriter);
#endif
    return font;
}

void setupAddressWidget(QLineEdit *widget, QWidget *parent)
{
    widget->setMaxLength(DarkSilkAddressValidator::MaxAddressLength);
    widget->setValidator(new DarkSilkAddressValidator(parent));
    widget->setFont(darksilkAddressFont());
}

void setupAmountWidget(QLineEdit *widget, QWidget *parent)
{
    QDoubleValidator *amountValidator = new QDoubleValidator(parent);
    amountValidator->setDecimals(8);
    amountValidator->setBottom(0.0);
    widget->setValidator(amountValidator);
    widget->setAlignment(Qt::AlignRight|Qt::AlignVCenter);
}

bool parseDarkSilkURI(const QUrl &uri, SendCoinsRecipient *out)
{
    // DarkSilk: check prefix
    if(uri.scheme() != QString("darksilk"))
        return false;

    SendCoinsRecipient rv;
    rv.address = uri.path();
    rv.amount = 0;
    QList<QPair<QString, QString> > items = uri.queryItems();
    for (QList<QPair<QString, QString> >::iterator i = items.begin(); i != items.end(); i++)
    {
        bool fShouldReturnFalse = false;
        if (i->first.startsWith("req-"))
        {
            i->first.remove(0, 4);
            fShouldReturnFalse = true;
        }

        if (i->first == "label")
        {
            rv.label = i->second;
            fShouldReturnFalse = false;
        }
        else if (i->first == "amount")
        {
            if(!i->second.isEmpty())
            {
                if(!DarkSilkUnits::parse(DarkSilkUnits::DRKSLK, i->second, &rv.amount))
                {
                    return false;
                }
            }
            fShouldReturnFalse = false;
        }

        if (fShouldReturnFalse)
            return false;
    }
    if(out)
    {
        *out = rv;
    }
    return true;
}

bool parseDarkSilkURI(QString uri, SendCoinsRecipient *out)
{
    // Convert darksilk:// to darksilk:
    //
    //    Cannot handle this later, because darksilk:// will cause Qt to see the part after // as host,
    //    which will lower-case it (and thus invalidate the address).
    if(uri.startsWith("darksilk://"))
    {
        uri.replace(0, 11, "darksilk:");
    }
    QUrl uriInstance(uri);
    return parseDarkSilkURI(uriInstance, out);
}

QString HtmlEscape(const QString& str, bool fMultiLine)
{
    QString escaped = Qt::escape(str);
    escaped = escaped.replace(" ", "&nbsp;");
    if(fMultiLine)
    {
        escaped = escaped.replace("\n", "<br>\n");
    }
    return escaped;
}

QString HtmlEscape(const std::string& str, bool fMultiLine)
{
    return HtmlEscape(QString::fromStdString(str), fMultiLine);
}

void copyEntryData(QAbstractItemView *view, int column, int role)
{
    if(!view || !view->selectionModel())
        return;
    QModelIndexList selection = view->selectionModel()->selectedRows(column);

    if(!selection.isEmpty())
    {
        // Copy first item
        QApplication::clipboard()->setText(selection.at(0).data(role).toString());
    }
}

QString getSaveFileName(QWidget *parent, const QString &caption,
                                 const QString &dir,
                                 const QString &filter,
                                 QString *selectedSuffixOut)
{
    QString selectedFilter;
    QString myDir;
    if(dir.isEmpty()) // Default to user documents location
    {
        myDir = QDesktopServices::storageLocation(QDesktopServices::DocumentsLocation);
    }
    else
    {
        myDir = dir;
    }
    QString result = QFileDialog::getSaveFileName(parent, caption, myDir, filter, &selectedFilter);

    /* Extract first suffix from filter pattern "Description (*.foo)" or "Description (*.foo *.bar ...) */
    QRegExp filter_re(".* \\(\\*\\.(.*)[ \\)]");
    QString selectedSuffix;
    if(filter_re.exactMatch(selectedFilter))
    {
        selectedSuffix = filter_re.cap(1);
    }

    /* Add suffix if needed */
    QFileInfo info(result);
    if(!result.isEmpty())
    {
        if(info.suffix().isEmpty() && !selectedSuffix.isEmpty())
        {
            /* No suffix specified, add selected suffix */
            if(!result.endsWith("."))
                result.append(".");
            result.append(selectedSuffix);
        }
    }

    /* Return selected suffix if asked to */
    if(selectedSuffixOut)
    {
        *selectedSuffixOut = selectedSuffix;
    }
    return result;
}

Qt::ConnectionType blockingGUIThreadConnection()
{
    if(QThread::currentThread() != QCoreApplication::instance()->thread())
    {
        return Qt::BlockingQueuedConnection;
    }
    else
    {
        return Qt::DirectConnection;
    }
}

bool checkPoint(const QPoint &p, const QWidget *w)
{
    QWidget *atW = qApp->widgetAt(w->mapToGlobal(p));
    if (!atW) return false;
    return atW->topLevelWidget() == w;
}

bool isObscured(QWidget *w)
{
    return !(checkPoint(QPoint(0, 0), w)
        && checkPoint(QPoint(w->width() - 1, 0), w)
        && checkPoint(QPoint(0, w->height() - 1), w)
        && checkPoint(QPoint(w->width() - 1, w->height() - 1), w)
        && checkPoint(QPoint(w->width() / 2, w->height() / 2), w));
}

void openDebugLogfile()
{
    boost::filesystem::path pathDebug = GetDataDir() / "debug.log";

    /* Open debug.log with the associated application */
    if (boost::filesystem::exists(pathDebug))
        QDesktopServices::openUrl(QUrl::fromLocalFile(QString::fromStdString(pathDebug.string())));
}

ToolTipToRichTextFilter::ToolTipToRichTextFilter(int size_threshold, QObject *parent) :
    QObject(parent), size_threshold(size_threshold)
{

}

bool ToolTipToRichTextFilter::eventFilter(QObject *obj, QEvent *evt)
{
    if(evt->type() == QEvent::ToolTipChange)
    {
        QWidget *widget = static_cast<QWidget*>(obj);
        QString tooltip = widget->toolTip();
        if(tooltip.size() > size_threshold && !tooltip.startsWith("<qt>") && !Qt::mightBeRichText(tooltip))
        {
            // Prefix <qt/> to make sure Qt detects this as rich text
            // Escape the current message as HTML and replace \n by <br>
            tooltip = "<qt>" + HtmlEscape(tooltip, true) + "<qt/>";
            widget->setToolTip(tooltip);
            return true;
        }
    }
    return QObject::eventFilter(obj, evt);
}

#ifdef WIN32
boost::filesystem::path static StartupShortcutPath()
{
    return GetSpecialFolderPath(CSIDL_STARTUP) / "DarkSilk.lnk";
}

bool GetStartOnSystemStartup()
{
    // check for DarkSilk.lnk
    return boost::filesystem::exists(StartupShortcutPath());
}

bool SetStartOnSystemStartup(bool fAutoStart)
{
    // If the shortcut exists already, remove it for updating
    boost::filesystem::remove(StartupShortcutPath());

    if (fAutoStart)
    {
        CoInitialize(NULL);

        // Get a pointer to the IShellLink interface.
        IShellLink* psl = NULL;
        HRESULT hres = CoCreateInstance(CLSID_ShellLink, NULL,
                                CLSCTX_INPROC_SERVER, IID_IShellLink,
                                reinterpret_cast<void**>(&psl));

        if (SUCCEEDED(hres))
        {
            // Get the current executable path
            TCHAR pszExePath[MAX_PATH];
            GetModuleFileName(NULL, pszExePath, sizeof(pszExePath));

            TCHAR pszArgs[5] = TEXT("-min");

            // Set the path to the shortcut target
            psl->SetPath(pszExePath);
            PathRemoveFileSpec(pszExePath);
            psl->SetWorkingDirectory(pszExePath);
            psl->SetShowCmd(SW_SHOWMINNOACTIVE);
            psl->SetArguments(pszArgs);

            // Query IShellLink for the IPersistFile interface for
            // saving the shortcut in persistent storage.
            IPersistFile* ppf = NULL;
            hres = psl->QueryInterface(IID_IPersistFile,
                                       reinterpret_cast<void**>(&ppf));
            if (SUCCEEDED(hres))
            {
                WCHAR pwsz[MAX_PATH];
                // Ensure that the string is ANSI.
                MultiByteToWideChar(CP_ACP, 0, StartupShortcutPath().string().c_str(), -1, pwsz, MAX_PATH);
                // Save the link by calling IPersistFile::Save.
                hres = ppf->Save(pwsz, TRUE);
                ppf->Release();
                psl->Release();
                CoUninitialize();
                return true;
            }
            psl->Release();
        }
        CoUninitialize();
        return false;
    }
    return true;
}

#elif defined(Q_OS_LINUX)

// Follow the Desktop Application Autostart Spec:
//  http://standards.freedesktop.org/autostart-spec/autostart-spec-latest.html

boost::filesystem::path static GetAutostartDir()
{
    namespace fs = boost::filesystem;

    char* pszConfigHome = getenv("XDG_CONFIG_HOME");
    if (pszConfigHome) return fs::path(pszConfigHome) / "autostart";
    char* pszHome = getenv("HOME");
    if (pszHome) return fs::path(pszHome) / ".config" / "autostart";
    return fs::path();
}

boost::filesystem::path static GetAutostartFilePath()
{
    return GetAutostartDir() / "darksilk.desktop";
}

bool GetStartOnSystemStartup()
{
    boost::filesystem::ifstream optionFile(GetAutostartFilePath());
    if (!optionFile.good())
        return false;
    // Scan through file for "Hidden=true":
    std::string line;
    while (!optionFile.eof())
    {
        getline(optionFile, line);
        if (line.find("Hidden") != std::string::npos &&
            line.find("true") != std::string::npos)
            return false;
    }
    optionFile.close();

    return true;
}

bool SetStartOnSystemStartup(bool fAutoStart)
{
    if (!fAutoStart)
        boost::filesystem::remove(GetAutostartFilePath());
    else
    {
        char pszExePath[MAX_PATH+1];
        memset(pszExePath, 0, sizeof(pszExePath));
        if (readlink("/proc/self/exe", pszExePath, sizeof(pszExePath)-1) == -1)
            return false;

        boost::filesystem::create_directories(GetAutostartDir());

        boost::filesystem::ofstream optionFile(GetAutostartFilePath(), std::ios_base::out|std::ios_base::trunc);
        if (!optionFile.good())
            return false;
        // Write a darksilk.desktop file to the autostart directory:
        optionFile << "[Desktop Entry]\n";
        optionFile << "Type=Application\n";
        optionFile << "Name=DarkSilk\n";
        optionFile << "Exec=" << pszExePath << " -min\n";
        optionFile << "Terminal=false\n";
        optionFile << "Hidden=false\n";
        optionFile.close();
    }
    return true;
}
#else

// TODO: OSX startup stuff; see:
// https://developer.apple.com/library/mac/#documentation/MacOSX/Conceptual/BPSystemStartup/Articles/CustomLogin.html

bool GetStartOnSystemStartup() { return false; }
bool SetStartOnSystemStartup(bool fAutoStart) { return false; }

#endif

HelpMessageBox::HelpMessageBox(QWidget *parent) :
    QMessageBox(parent)
{
    header = tr("DarkSilk-Qt") + " " + tr("version") + " " +
        QString::fromStdString(FormatFullVersion()) + "\n\n" +
        tr("Usage:") + "\n" +
        "  darksilk-qt [" + tr("command-line options") + "]                     " + "\n";

    coreOptions = QString::fromStdString(HelpMessage());

    uiOptions = tr("UI options") + ":\n" +
        "  -lang=<lang>           " + tr("Set language, for example \"de_DE\" (default: system locale)") + "\n" +
        "  -min                   " + tr("Start minimized") + "\n" +
        "  -splash                " + tr("Show splash screen on startup (default: 1)") + "\n";

    setWindowTitle(tr("DarkSilk-Qt"));
    setTextFormat(Qt::PlainText);
    // setMinimumWidth is ignored for QMessageBox so put in non-breaking spaces to make it wider.
    setText(header + QString(QChar(0x2003)).repeated(50));
    setDetailedText(coreOptions + "\n" + uiOptions);
}

void HelpMessageBox::printToConsole()
{
    // On other operating systems, the expected action is to print the message to the console.
    QString strUsage = header + "\n" + coreOptions + "\n" + uiOptions;
    fprintf(stdout, "%s", strUsage.toStdString().c_str());
}

void HelpMessageBox::showOrPrint()
{
#if defined(WIN32)
        // On Windows, show a message box, as there is no stderr/stdout in windowed applications
        exec();
#else
        // On other operating systems, print help text to console
        printToConsole();
#endif
}

ClickableLabel::ClickableLabel( const QString& text, QWidget * parent ) :
    QLabel(parent)

{
    this->setText(text);
}

ClickableLabel::~ClickableLabel()
{
}

void ClickableLabel::mouseReleaseEvent ( QMouseEvent * event )
{
    emit clicked();
}

void SetBlackThemeQSS(QApplication& app)
{
    app.setStyleSheet("QTableView { background: rgb(0,0,0); color: rgb(255,255,255); border: 1px inset white; border-radius: 3px; margin: 0px; }"
                      "QWidget        { background-color: rgb(0,0,0); alternate-background-color: rgb(0,0,0); color: rgb(255,255,255);}"
                      "QFrame         { border: none; }"
                      "QCheckBox      { color: rgb(0,0,0); background-color: rgb(255,255,255);}"
                      "QComboBox      { background-color: rgb(0,0,0); color: rgb(255,255,255); }"
                      "QComboBox QAbstractItemView::item { color: rgb(255,255,255); }"
                      "QPushButton    { background: rgb(86,0,120); alternate-background-color: rgb(86,0,120); color: rgb(255,255,255); }"
                      "QDoubleSpinBox { background: rgb(0,0,0); color: rgb(255,255,255); border-color: rgb(255,0,0); }"
                      "QLineEdit      { background: rgb(0,0,0); color: rgb(255,255,255); border-color: rgb(255,0,0); }"
                      "QTextEdit      { background: rgb(0,0,0); color: rgb(255,255,255); border: 1px inset white; border-radius: 3px; margin: 0px; }"
                      "QPlainTextEdit { background: rgb(0,0,0); color: rgb(255,255,255); }"
                      "QMenuBar       { background: rgb(0,0,0); alternate-background-color: rgb(0,0,0); color: rgb(255,255,255); border-color: rgb(0,0,0); }"
                      "QMenuBar::item {background-color: rgb(0,0,0); }"
                      "QMenu::item:selected { background-color: rgb(86,0,120); }"
                      "QLabel         { color: rgb(255,255,255); }"
                      "QScrollBar     { color: rgb(255,255,255); }"
                      "QCheckBox      { color: rgb(0,0,0); background-color: rgb(255,255,255); }"
                      "QRadioButton   { color: rgb(0,0,0); }"
                      "QTabBar::tab   { color: rgb(255,255,255); border: 1px solid rgb(255,255,255); border-bottom: none; padding: 5px; }"
                      "QTabBar::tab:selected  { background-color: rgb(86,0,120); }"
                      "QTabBar::tab:!selected { background-color: rgb(0,0,0); margin-top: 2px; }"
                      "QTabWidget::pane { border: 1px solid rgb(0,0,0); }"
                      "QToolButton    { background: rgb(0,0,0); color: rgb(232,234,236); border: none; border-left-color: rgb(0,0,0); border-left-style: solid; border-left-width: 6px; margin-top: 8px; margin-bottom: 8px; }"
                      "QToolButton:checked { color: rgb(255,255,255); border: none; border-left-color: rgb(86,0,120); border-left-style: solid; border-left-width: 6px; }"
                      "QProgressBar   { color: rgb(255,255,255); border: 1px inset white; border-radius: 3px; margin: 0px; }"
                      "QProgressBar::chunk { background: rgb(86,0,120); }"
                      "QTreeView::item { background: rgb(86,0,120); color: rgb(255,255,255); }"
                      "QTreeView::item:selected { background-color: rgb(138,138,138); }"
                      "QHeaderView::section { background-color: rgb(0,0,0); color: rgb(255,255,255); }"
                      "QToolBar       { background: rgb(0,0,0); border: rgb(0,0,0); }"
                      "QTextEdit      { background-color: rgb(0,0,0); color: rgb(255,255,255); }"
                      "QPlainTextEdit { background-color: rgb(0,0,0); color: rgb(255,255,255); }"
                      "QProgressBar { background-color: rgb(0,0,0); border: 1px solid grey; border-radius: 7px; padding: 1px; text-align: center; } QProgressBar::chunk { background: QLinearGradient(x1: 0, y1: 0, x2: 1, y2: 0, stop: 0 #220050, stop: 1 #500078); border-radius: 7px; margin: 0px; }"
                      );
}

QString formatPingTime(double dPingTime)
{
    return dPingTime == 0 ? QObject::tr("N/A") : QString(QObject::tr("%1 ms")).arg(QString::number((int)(dPingTime * 1000), 10));
}

QString formatDurationStr(int secs)
{
    QStringList strList;
    int days = secs / 86400;
    int hours = (secs % 86400) / 3600;
    int mins = (secs % 3600) / 60;
    int seconds = secs % 60;

    if (days)
        strList.append(QString(QObject::tr("%1 d")).arg(days));
    if (hours)
        strList.append(QString(QObject::tr("%1 h")).arg(hours));
    if (mins)
        strList.append(QString(QObject::tr("%1 m")).arg(mins));
    if (seconds || (!days && !hours && !mins))
        strList.append(QString(QObject::tr("%1 s")).arg(seconds));

    return strList.join(" ");
}

QString formatServicesStr(quint64 mask)
{
    if (mask == (NODE_I2P | NODE_NETWORK))
        return "P2P Nodes and I2P";

    if (mask == NODE_NETWORK)
        return "P2P Nodes";

    return "Unknown";

}

} // namespace GUIUtil

